src/linear_algebra/unitary_group.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Shing Tak Lam
+-/
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import data.complex.basic
+
+/-!
+# The Unitary Group
+
+This file defines elements of the unitary group `unitary_group n α`, where `α` is a `star_ring`.
+This consists of all `n` by `n` matrices with entries in `α` such that the star-transpose is its
+inverse. In addition, we define the group structure on `unitary_group n α`, and the embedding into
+the general linear group `general_linear_group α (n → α)`.
+
+We also define the orthogonal group `orthogonal_group n β`, where `β` is a `comm_ring`.
+
+## Main Definitions
+
+ * `matrix.unitary_group` is the type of matrices where the star-transpose is the inverse
+ * `matrix.unitary_group.group` is the group structure (under multiplication)
+ * `matrix.unitary_group.embedding_GL` is the embedding `unitary_group n α → GLₙ(α)`
+ * `matrix.orthogonal_group` is the type of matrices where the transpose is the inverse
+
+## References
+
+ * https://en.wikipedia.org/wiki/Unitary_group
+
+## Tags
+
+matrix group, group, unitary group, orthogonal group
+
+-/
+
+namespace matrix
+universes u v
+open linear_map
+open_locale matrix
+
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n]
+variables (α : Type v) [comm_ring α] [star_ring α]
+
+/--
+`unitary_submonoid n α` is the submonoid consisting of all `n` by `n` matrices `A` where
+`star A ⬝ A = 1`.
+-/
+def unitary_submonoid : submonoid (matrix n n α) :=
+{ carrier := {A | star A ⬝ A = 1},
+  one_mem' := show star 1 ⬝ 1 = 1, by { rw [star_one, matrix.one_mul] },
+  mul_mem' := λ A B (hA : star A ⬝ A = 1) (hB : star B ⬝ B = 1), show star (A ⬝ B) ⬝ (A ⬝ B) = 1,
+    by { rwa [star_mul, ←matrix.mul_assoc, matrix.mul_assoc _ _ A, hA, matrix.mul_one] } }
+
+/--
+`unitary_group n` is the group of `n` by `n` matrices where the star-transpose is the inverse.
+-/
+@[derive monoid]
+def unitary_group : Type* := unitary_submonoid n α
+
+end
+
+variables {α : Type v} [comm_ring α] [star_ring α]
+variables {n : Type u} [decidable_eq n] [fintype n]
+
+namespace unitary_submonoid
+
+lemma star_mem {A : matrix n n α} (h : A ∈ unitary_submonoid n α) :
+  star A ∈ unitary_submonoid n α :=
+matrix.nonsing_inv_left_right _ _ $ (star_star A).symm ▸ h
```

Comment:
In particular, this lemma can only be proven for matrices for now, until we generalize `nonsing_inv`

Reply:
should I add a TODO here? I don't know enough of the maths to say how this should be generalised. In particular, if we have `unitary_submonoid M`), I think if `M` is a commutative monoid this is trivial, and if `M` is finite then this is also true. But I don't know how this generalises to arbitrary monoids, as left inverses aren't guaranteed to be right inverses.